[MRG] Sample weights for ElasticNet

lorentzenchr opened the pull request:
#### Reference Issues/PRs
Partially solves #3702: Adds `sample_weight` to `ElasticNet` and `Lasso`, but only for dense feature array `X`.

Commit:
MNT Simplify _check_sample_weight

Commit:
TST Fix unsupported sample weights

Commit:
ENH Support sample weights for dense ElasticNet

Commit:
TST raise ValueError when sparse enet with sample weights

Commit:
Fix sparse logic in _rescale_data

Commit:
Remove check for negative sample weights

Commit:
TST test _rescale_data_ for sparse input

Commit:
Merge branch 'master' into enet_sample_weights

rth commented on sklearn/compose/tests/test_target.py:
```diff
@@ -37,7 +37,7 @@ def test_transform_target_regressor_error():
     # fit with sample_weight with a regressor which does not support it
     sample_weight = np.ones((y.shape[0],))
-    regr = TransformedTargetRegressor(regressor=Lasso(),
+    regr = TransformedTargetRegressor(regressor=OrthogonalMatchingPursuit(),
```
Why this change?

lorentzenchr replied:
Here, one needs an estimator that does not support `sample_weight`. As `Lasso` inherits sample_weight support from `ElasticNet`, I needed to change this.

rth commented on sklearn/linear_model/_base.py:
```diff
@@ -191,6 +195,26 @@ def _rescale_data(X, y, sample_weight):
+        if order == 'F':
+            X = sparse.csc_matrix(X)
+        else:
+            X = sparse.csr_matrix(X)
```
Maybe,
```py
sparse_format = "csc" if order == "F" else "csr"

X = X.asformat(sparse_format, copy=False)
```

that should avoid copies if the sparse matrix is already in the right format..

rth commented on sklearn/linear_model/_base.py:
```diff
@@ -191,6 +195,26 @@ def _rescale_data(X, y, sample_weight):
+    elif order == 'F':
+        X = np.asfortranarray(X)
+    else:
+        X = np.ascontiguousarray(X)
```
Maybe just,
```py
X = np.asarray(X, order=order)
```
after having checked that `order is in ['C', 'F'].

rth commented on sklearn/linear_model/_base.py:
```diff
@@ -179,9 +180,12 @@ def _preprocess_data(X, y, fit_intercept, normalize=False, copy=True,
 # sample_weight makes the refactoring tricky.
 
-def _rescale_data(X, y, sample_weight):
+def _rescale_data(X, y, sample_weight, order='C'):
```
Do we actually always need the order? I guess some of the algorithms work either way. Maybe add keep `order=None` by default which wouldn't constrain order.

rth commented on sklearn/linear_model/_base.py:
```diff
@@ -179,9 +180,12 @@ def _preprocess_data(X, y, fit_intercept, normalize=False, copy=True,
+    sparse_X = sparse.issparse(X)
+    sparse_y = sparse.issparse(y)
+    sample_weight = np.asarray(sample_weight)
     sample_weight = np.asarray(sample_weight)
```
Redudnant line

rth commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -703,18 +707,44 @@ def fit(self, X, y, check_input=True):
                      self.fit_intercept, copy=should_copy,
-                     check_input=check_input)
+                     check_input=check_input, sample_weight=sample_weight,
+                     order='F')
```
Maybe worth adding a comment about why `order="F"` is more efficient here

lorentzenchr replied:
It's only more efficient code at this place, not better performance.

rth commented on sklearn/linear_model/tests/test_base.py:
```diff
@@ -433,20 +433,49 @@ def test_dtype_preprocess_data():
+        assert rescaled_y.flags['C_CONTIGUOUS']
+    elif order == 'F':
+        assert rescaled_X.flags['F_CONTIGUOUS']
+        assert rescaled_y.flags['F_CONTIGUOUS']
```
In most cases, y is a 1D array so it's F and C continuous anyway..

lorentzenchr replied:
I added tests for 2d target y.

rth reviewed (commented):
Thanks for working on this @lorentzenchr ! Haven't done a full review, but a few comments are below.

Commit:
DOC align parameter doc of sample_weight

Commit:
DOC adds docstring to _rescale_data

Commit:
Simplify _rescale_data with reviewer suggestion

Commit:
TST invalid order argument in _rescale_data

Commit:
add doc and comment on order='F'

Commit:
TST 2d target in _rescale_data

Commit:
Unsupported arg copy in asformat prior to scipy 1.1.0

Commit:
Merge branch 'master' into enet_sample_weights

Commit:
Merge branch 'master' into enet_sample_weights

Commit:
TST test sample_weight=number same as sw=None

Commit:
Merge branch 'master' into enet_sample_weights

rth commented on sklearn/linear_model/_base.py:
```diff
@@ -191,6 +211,20 @@ def _rescale_data(X, y, sample_weight):
     y = safe_sparse_dot(sw_matrix, y)
+
+    if order is not None:
+        sparse_format = "csc" if order == "F" else "csr"
```
I find this somewhat not obvious, when calling `_rescale_data(..., order='C')` on a sparse array. In fact this doesn't need to be part of `_rescale_data` it could be a separate private function enforcing order when needed.

lorentzenchr replied:
I moved this in a private function _set_order in `__coordinate_descent.py`. It's the only place where it is used.

rth reviewed (commented):
So this adds support for sample weight by rescaling X by sqrt(sample_weight). I know that it is done in other linear model, but how can we be sure that this is the right thing to do here with L1 regularization?

It would have been nice to have a test comparing results using sample_weight with Ridge when `l1_ratio=0`. For  `l1_ratio > 0` do you have any suggestions for checking in unit tests that the output is correct (beyond being consistent with respect to expected invariances) ?

For all the added conversions for array order and sparse format, those are optimizations orthogonal to adding sample_weights, right? If it might be better to separate that code into a separate private function, used when necessary with some justification that this does impact performance in a good way (it's hard to judge which order is optimal without actually benchmarking it for me). This part could also be a separate PR.

lorentzenchr commented:
@rth You raise some good questions.

> how can we be sure that this is the right thing to do here with L1 regularization?

Because rescaling by the square root works and I thought it is the minimal code change. If I were to code it from scratch, I would include sample weights in the loss function, it's gradient and hessian and pass it to the optimizer.

>It would have been nice to have a test comparing results using sample_weight with Ridge when `l1_ratio=0`.

I will try to include such a test.

>For l1_ratio > 0 do you have any suggestions for checking in unit tests that the output is correct?

One can check against [glmnet](https://cran.r-project.org/web/packages/glmnet). What do you think?

>For all the added conversions for array order and sparse format, those are optimizations orthogonal to adding sample_weights, right?

The only solver for L1-penalty in scikit-learn is coordinate descent (see also #12966). It is written in Cython and expects F-ordered arrays which is optimal for this algo. All I did was to avoid memory copies and pass F-ordered arrays to `_cd_fast`. Though I might have missed something and I'm not an expert on memoryviews in Cython.
Therefore, I would prefer to keep this PR together.

rth commented:
>> how can we be sure that this is the right thing to do here with L1 regularization?
>
> Because rescaling by the square root works and I thought it is the minimal code change.

Indeed. Actually, if we take as definition of sample weights that a weight of 2 is equivalent to 2 repeated samples etc, https://github.com/scikit-learn/scikit-learn/issues/15651#issuecomment-555210612 then checking this invariances as you have done could be enough, and I'm actually fine with this without a comparison to glmnet or other external package. Other parameters (alpha, l1_ratio) should be invariant to the addition of sample weights, as far as I understand (to come back to https://github.com/scikit-learn/scikit-learn/issues/15657#issuecomment-555747648 discussion).

Also cc @agramfort 

> All I did was to avoid memory copies and pass F-ordered arrays to _cd_fast.
> Therefore, I would prefer to keep this PR together.

OK, fair enough.

agramfort commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -897,3 +898,94 @@ def test_multi_task_lasso_cv_dtype():
+        assert_allclose(reg.intercept_, intercept1)
+
+    # check that multiplying sample_weight by 2 is equivalent
+    # to repeating correspoding samples twice
```
```suggestion
    # to repeating corresponding samples twice
```

agramfort replied:
do you confirm that this not a systematic property of sample weights but
that it works here as we have 1/n_samples in front of the loss for Lasso and ElasticNet?
For our parametrization of Ridge or SVC/SVR it would not work right?

rth replied:
> do you confirm that this not a systematic property of sample weights but
that it works here as we have 1/n_samples in front of the loss for Lasso and ElasticNet?
For our parametrization of Ridge or SVC/SVR it would not work right?

Yes, as far as I understand. For Ridge one would also need to adjust alpha to obtain equivalent result.

lorentzenchr replied:
Yes, I can confirm that. It is as @rth explained.

agramfort commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -709,18 +713,46 @@ def fit(self, X, y, check_input=True):
+            if check_input:
+                if sparse.issparse(X):
+                    raise ValueError("Sample weights do not (yet) support "
+                                     "sparse matrices.")
```
I see tests with sample weights and sparse matrices why do you have this?

lorentzenchr replied:
There is `test_enet_sample_weight_sparse`, whicht tests that `ElasticNetc` raises an error as sample weights are not supported with sparse `X`.
Secondly, there is `test_enet_sample_weight_consistency` which is decorated with `@pytest.mark.parametrize('sparseX', [False])` such that this test could easily be extended once sparse `X` are supported with sample weights.

agramfort replied:
ok I had not paid attention to `@pytest.mark.parametrize('sparseX', [False])` We tend to add code for tests when it becomes relevant. Do you plan to this next when this is merged?

lorentzenchr replied:
I'd say that this is needed to close #3702 and that is is a useful feature. I'm tempted to try it in a new PR, but I'm afraid that it needs a larger code change than this one.

agramfort commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -897,3 +898,94 @@ def test_multi_task_lasso_cv_dtype():
+        assert_allclose(reg.intercept_, intercept)
+
+    # setting one element of sample_weight to 0 is equivalent to removing
+    # the correspoding sample
```
```suggestion
    # the corresponding sample
``

Commit:
DOC fix a typo

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

Commit:
DOC correct a typo

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

Commit:
CLN remove unwanted line

agramfort commented:
2 more things on my side. You have left comments from @rth to address or rule out. I would not add code/test for things we don't use and we don't know if the new code is actually correct. I would therefore remove the sparseX option in the test. Let's had code / tests when we actually use it. thx @lorentzenchr

Commit:
MAINT use _astype_copy_false

Commit:
RFC private _set_order function

Commit:
TST remove sparse in enet with sample weights

lorentzenchr commented:
@agramfort Thanks for pointing out unaddressed issues. I hope it is in a good shape now.
I suspect and hope that the test error of doc-min-dependencies in unrelated to this PR.

agramfort reviewed (approved):
LGTM

@rth feel free to have another look

thx @lorentzenchr

agramfort commented:
@lorentzenchr just don't forget to update what's new page.

Commit:
Merge branch 'master' into enet_sample_weights

Commit:
DOC what's new

agramfort commented on doc/whats_new/v0.23.rst:
```diff
@@ -154,10 +154,14 @@ Changelog
 ...........................
 
+- |Feature| Support of `sample_weight` in :class:`linear_model.ElasticNet` and
+  :class:`linear_model:Lasso` for dense feature matrix `X`.
```
you don't have it for the CV estimators?

lorentzenchr replied:
No. The what's new is correct. Sample weight support is added only to `ElasticNet` and `Lasso`. Unfortunately, the logic in `_coordinate_descent.py` is such that this feature is not automatically inherited by `ElasticNetCV` and others. I do not know if there are advantages with the current structure.
In any case, I'd prefer to put adding sample weights to ElasticNetCV in another PR. What do you think?

agramfort commented:
ok. It's just that I hear something that we are consistent in linear models
with certain options implemented
only in certain estimators without clear reason. So it suggests we should
have sample_weights added to CV
classes very soon and ideally in the same release cycle

>

rth reviewed (approved):
LGTM as well, thanks @lorentzenchr !
